pkg/reconcile/reconcile.go

```diff
@@ -0,0 +1,129 @@
+package reconcile
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// DiffFunc takes two Kubernetes resources: expected and existing. Both are assumed to be
+// the same Kind. It compares the two, and returns an Operation indicating how to
+// transition from existing to expected. If an update is required, it will set the
+// relevant fields on existing to their intended values. This is so that we can simply
+// resubmit the existing resource, and any fields automatically set by the Kubernetes API
+// server will be retained.
+type DiffFunc func(runtime.Object, runtime.Object) Operation
+
+// Operation describes the operation performed by CreateOrUpdate.
+type Operation string
+
+var (
+	Create   Operation = "create"
+	Recreate Operation = "recreate"
+	Update   Operation = "update"
+	None     Operation = "none"
+	Error    Operation = "error"
+)
+
+// ObjWithMeta describes a Kubernetes resource with a metadata field. It's a combination
+// of two common existing Kubernetes interfaces. We do this because we want to use methods
+// from each in CreateOrUpdate, whilst still keeping the argument type generic.
+type ObjWithMeta interface {
+	metav1.Object
+	runtime.Object
+}
+
+// CreateOrUpdate takes a Kubernetes object and a "diff function" and attempts to ensure
+// that the the object exists in the cluster with the correct state. It will use the diff
+// function to determine any differences between the cluster state and the local state and
+// use that to decide how to update it.
+func CreateOrUpdate(ctx context.Context, c client.Client, existing ObjWithMeta, kind string, diffFunc DiffFunc) (Operation, error) {
+	name := types.NamespacedName{
+		Namespace: existing.GetNamespace(),
+		Name:      existing.GetName(),
+	}
+	expected := existing.(runtime.Object).DeepCopyObject()
+	err := c.Get(ctx, name, existing)
+	if err != nil {
+		if !apierror.IsNotFound(err) {
+			return Error, err
+		}
+		if err := c.Create(ctx, existing); err != nil {
+			return Error, err
+		}
+		return Create, nil
+	}
+
+	// existing contains the state we just fetched from Kubernetes.
+	// expected contains the state we're trying to reconcile towards.
+	// If an update is required, DiffFunc will set the relevant fields on existing such that we
+	// can just resubmit it to the cluster to achieve our desired state.
```

Comment:
The naming in this function is quite carefully chosen, and there's a subtle interaction between a couple of things here (all caused by pointers).

If you were to rename these variables such that we passed in an `expected` resource, you'd find that a load of tests fail. This is due to two things:
- We try to keep any existing state that Kube has set on the resource by fetching it, setting fields on it, and updating it
- The caller expects that, after calling `CreateOrUpdate`, the resource it passed it (as a pointer) will contain the most up-to-date state from Kube.

This means we must submit the exact resource passed in to Kube - not the copy we make for comparison purposes.

We could return the up-to-date resource from the function, and that would get around this, but I think we'd then need to type cast it back to the specific resource type, as the only type we could put in this function signature is `ObjectWithMeta`.